cranelift/interpreter/src/interpreter.rs

```diff
@@ -310,23 +371,62 @@ impl<'a> State<'a, DataValue> for InterpreterState<'a> {
         Address::from_parts(size, AddressRegion::Stack, 0, final_offset)
     }
 
-    fn heap_address(&self, _size: AddressSize, _offset: u64) -> Result<Address, MemoryError> {
-        unimplemented!()
+    fn heap_address(
+        &self,
+        size: AddressSize,
+        heap: Heap,
+        offset: u64,
+    ) -> Result<Address, MemoryError> {
+        // It is possible for this function to be called before the interpreter has started
+        // execution (see `InterpreterState::get_heap_address`). If this happens we cannot
+        // perform global value resolution, and have to build an address directly referencing
+        // the `heap` that was provided.
+        let has_current_function = self.frame_stack.len() == 0;
+        let addr = if has_current_function {
+            let heap_id = heap.as_u32() as u64;
+            Address::from_parts(size, AddressRegion::Heap, heap_id, offset)?
+        } else {
+            let heap_data = &self.get_current_function().heaps[heap];
+            let heap_base = self.resolve_global_value(heap_data.base)?;
+            let mut addr = Address::try_from(heap_base)?;
+            addr.size = size;
+            addr.offset += offset;
```

Comment:
> the heap base is itself an address with a heap and offset component?
Well, yes it could be. Resolving global values means that that address could be anywhere in memory, even in the stack. 

> In other words, do heaps resolve into regions within other heaps?
Yes, we can have multiple CLIF Heaps that resolve to the same heap in the interpreter.

This function was quite a bit confusing because we used the same `Heap` type to mean both a CLIF Heap and a Interpreter Heap. I've now separated those two into separate types, and the function now
only does CLIF Heap resolution.